src/test/java/hudson/plugins/jacoco/e2e/E2ETest.java

```diff
@@ -0,0 +1,108 @@
+package hudson.plugins.jacoco.e2e;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeDiagnosingMatcher;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.RealJenkinsRule;
+
+import hudson.Functions;
+import hudson.model.Descriptor;
+import hudson.model.FreeStyleBuild;
+import hudson.model.FreeStyleProject;
+import hudson.plugins.jacoco.JacocoBuildAction;
+import hudson.plugins.jacoco.JacocoPublisher;
+import hudson.plugins.jacoco.model.Coverage;
+import hudson.tasks.BatchFile;
+import hudson.tasks.Builder;
+import hudson.tasks.Publisher;
+import hudson.tasks.Shell;
+import hudson.util.DescribableList;
+
+import static org.hamcrest.CoreMatchers.*;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+import static hudson.plugins.jacoco.e2e.E2ETest.CoverageMatcher.withCoverage;
+
+public class E2ETest {
+
+    @Rule
+    public RealJenkinsRule rjr = new RealJenkinsRule();
+
+    @Test
+    public void simpleTest() throws Throwable {
+        rjr.then(r -> {
+            FreeStyleProject project = r.createFreeStyleProject();
+            project.getBuildersList().addAll(createJacocoProjectBuilders());
+            project.getPublishersList().add(new JacocoPublisher());
+            
+            FreeStyleBuild build = r.buildAndAssertSuccess(project);
+
+            assertThat("plugin collected data", build.getLog(), containsString("Collecting JaCoCo coverage data"));
+
+            JacocoBuildAction action = build.getAction(JacocoBuildAction.class);
+            assertThat("Build has the Jacoco Action", action, notNullValue());
+            
+            assertThat("incorrect branch coverage reported", action.getBranchCoverage(), withCoverage(0, 621, 621));
+            assertThat("incorrect class coverage reported", action.getClassCoverage(), withCoverage(7, 59, 66));
+            assertThat("incorrect complexity coverage reported", action.getComplexityScore(), withCoverage(19, 835, 854));
+            assertThat("incorrect instruction coverage reported", action.getInstructionCoverage(), withCoverage(229, 9013, 9242));
+            assertThat("incorrect line coverage reported", action.getLineCoverage(), withCoverage(53, 1860, 1913));
+        }
+        );
+    }
+
+    private static List<Builder> createJacocoProjectBuilders() {
+        String[] commands = { "git clone --branch jacoco-3.2.0 https://github.com/jenkinsci/jacoco-plugin.git .",
```

Comment:
whilst running maven inside a test is not the best, we are using this project as the test (and a tag) so we at least have most of the dependencies already local.

It will diverge over time, so we may want to try and keep this somewhat in sync

Reply:
So binaries "git" and "mvn" need to be available on the CI node for this to work, right?

Reply:
yes,  but maven already is as it is what runs the tests.

as for `git` the CI could have cloned using `jgit` rather than the `git` CLI so it may not be available, but ci.jenkins.io is not setup like that.   
A developer could work exclusively with IntelliJ or Eclipse (again `jgit`) but I am not convinced that is likely.

Reply:
For a more reproducible environment:

* store a small test project in `src/test/resources/sampleproj/`
* use `dependency:copy` to unpack a specific release of Maven into `target/test-classes/`

but still use the Java in path I guess.